include/boost/lockfree/spsc_queue.hpp

```diff
@@ -761,6 +856,41 @@ class spsc_queue:
         return base_type::push(t);
     }
 
+#ifndef BOOST_NO_CXX11_RVALUE_REFERENCES
+
+    /** Pushes object t to the ringbuffer via move construction/
+     *
+     * \pre only one thread is allowed to push data to the spsc_queue
+     * \post object will be pushed to the spsc_queue, unless it is full.
+     * \return true, if the push operation is successful.
+     *
+     * \note Thread-safe and wait-free
+     * */
+
+    bool push(T&& t)
+    {
+        return base_type::push(std::move(t));
+    }
+
+#ifndef BOOST_NO_CXX11_VARIADIC_TEMPLATES
+    /** Emplaces an instance of T to the ringbuffer via direct initialization using the given constructor arguments
+     *  
+     * \pre only one thread is allowed to push data to the spsc_queue
+     * \post object will be pushed to the spsc_queue, unless it is full.
+     * \return true, if the push operation is successful.
+     *
+     * \note Thread-safe and wait-free
+     * */
+    template<typename... Args>
+    typename boost::enable_if< typename boost::is_constructible<T, Args...>::type, bool>::type
+    emplace(Args&&... args)
+    {
+        return  base_type::emplace(std::forward<Args>(args)...);
+    }
+#endif
+
+#endif
+
     /** Pops one object from ringbuffer.
      *
      * \pre only one thread is allowed to pop data to the spsc_queue
```

Comment:
Just dreaming here, but:

It may be worth considering converting this argument-less `pop()` from returning `bool` to `boost::optional<T>`, which is convertible to bool. Because the current version, just returning `bool` is kind of worthless for everything except "some event has happened", similar to `boost::optional<void>` (if we had [regular void](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2015/p0146r0.html)). That would allow user code like this:
```cpp
while (auto item = queue.pop())
{
  // process item
}
```

Reply:
true. one could only introduce a new API `queue.pop_optional()`. i'd merge a PR if you implement it :)